demoapp/demo_app_functions.py

```diff
@@ -450,4 +466,479 @@ def update_csv(bill_num: str, title: str, summarized_bill: str, category: str, t
+        return 1, call_llm(bill_details, query, llm_call_type, return_metrics)
+
+    elif task == 'tag':
+
+        query_1 = get_query_for_categorizing(bill_details, llm_call_type)
+        category_response = call_llm(bill_details, query_1, llm_call_type, return_metrics)
+        categories = extract_categories_tags(category_response.response)
+        category_tags = get_category_tags(categories)
+        query_2 = get_query_for_tagging(bill_details, category_tags, llm_call_type)
+        tag_response = call_llm(bill_details, query_2, llm_call_type, return_metrics)
+        tag_response.response = extract_categories_tags(tag_response.response)
+        return 1, tag_response
+
+    return -2, LLMResults()
+
+def extract_categories_tags(response: str) -> list:     
+
+    '''
+    This function parses the response generated by LLM and extracts either categories or tags.
+
+    Args: 
+        response (str): Response string generated by the LLM. Ex: "Tags: ['tag_1', 'tag_2']"
+
+    Returns: 
+        list of either tags or categories. Ex: [tag_1, tag_2]
+
+    '''
+    
+    cats_str = response.split(':')[1]
+    
+    translate_dict = dict.fromkeys(map(ord, '-[]"'), None)
+    translate_dict[ord("'")] = None
+    cats_str = cats_str.translate(translate_dict)
+
+    return [i.strip() for i in cats_str.split(',')]
+
+def prepare_invoke_dict(bill_details: BillDetails) -> dict:
+
+    """
+    This function prepares the dict object that is used in chain.invoke function to call the LLM with prompt and 
+    required details. 
+
+    Args: 
+        bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+
+    Returns: 
+        dict object containing all the necessary keys and values required for invoke call. 
+
+    """
+    return {
+                "title": bill_details.bill_title, 
+                "context": [Document(page_content=x) for x in bill_details.bill_text],
+                "names": bill_details.mgl_names, 
+                "mgl_sections": [Document(page_content=x) for x in bill_details.mgl_ref],
+                "committee_info": bill_details.committee_info
+            }
+
+def get_query_for_summarization(bill_details: BillDetails, llm_call_type: str) -> str:
+
+    """
+        
+    This functions prepares a prompt based on the call type (small: No use of RAG, large: Use RAG) for bill summarization
+
+    Args: 
+        bill_details (BillDetails): Object containing information about the bill - bill_text, bill_title, mgl_ref, commottee_info, mgl_names
+        llm_call_type (str): This argument can take 2 values ("small": No use of RAG, "large": Use RAG) 
+
+    Returns: 
+        str: Query that includes task, context and instructions on summary generation
+
+    """
+
+    
+    if llm_call_type == 'large':
+        query = f""" Can you please explain what the following MA bill means to a regular resident without specialized knowledge? 
```

Comment:
When you move this into a new file, please abstract the query text into a toplevel (module-level) constant. That will allow us to easily modify it or even import it into other functions.